pytorch_lightning/callbacks/stochastic_weight_avg.py

```diff
@@ -159,7 +169,9 @@ def on_fit_start(self, trainer: "pl.Trainer", pl_module: "pl.LightningModule"):
             trainer.fit_loop.max_epochs += 1
 
     def on_train_epoch_start(self, trainer: "pl.Trainer", pl_module: "pl.LightningModule"):
-        if trainer.current_epoch == self.swa_start:
+        if (not self._initialized) and (self.swa_start <= trainer.current_epoch <= self.swa_end):
```

Comment:
okay... as I anticipated the fix is not complete. When it resumes the training:
- the checkpoint contains the scheduler states as well.
- Now if in the previous run, SWALR gets configured, then the scheduler states, will be of SWALR.
- While resuming in such a condition, when we reload the scheduler states, SWALR is not configured by that point and SWALR scheduler states will be loaded inside another scheduler configured by the user explicitly which might reload the states sometimes (as it does in the test) but might not reload the state if some other incompatible scheduler is configured (as in https://github.com/PyTorchLightning/pytorch-lightning/issues/11665 issue)

need to think more on this.

Reply:
I've pushed a commit with a unit test that reproduces this crash to a new branch: https://github.com/adamreeve/pytorch-lightning/commit/33f67eb7b918468e28b32805f0112e375d6f6dc5

Not sure yet how to handle this though. I've tried configuring the scheduler within `StochasticWeightAveraging.on_load_checkpoint` but that doesn't work as the optimizers haven't yet been configured at the point the callback state is restored, and that only happens right before restoring the scheduler state.

If we could get the correct SWALR scheduler configured before its checkpoint state is loaded that would probably mean I don't need my somewhat hacky saving and restoring of the scheduler's `_step_count`.

Reply:
Okay I have a fix for this, although it's still on a separate branch for now: https://github.com/adamreeve/pytorch-lightning/commit/0107ff1205ed39509f0487ca5039899f64b36428

It feels a little bit flaky as it relies on the order in which checkpoint state is restored. The idea is to hide the scheduler config so that the CheckpointConnector doesn't try to restore state for them, then restore the configs with the scheduler state from the checkpoint when training restarts.

If this seems reasonable I can merge it into this PR.

Reply:
I've merged those changes into this PR now